weather API has been added to Account page

InnaKamenskaya opened the pull request:
API for displaying weather data received from an external service was implemented in two ways. From front side by using uiRecordApi to get information about city and apiKey for authorization to external service. From back side by using Apex HTTP Callouts. To switch between realization, please use Is Back Implementation checkbox on the Detail page of the Account.

InnaKamenskaya committed:
weather API has been added to Account page

okrasko55 commented on force-app/main/default/classes/WeatherService.cls:
```diff
@@ -0,0 +1,33 @@
+    @AuraEnabled(cacheable=true)
+    public static WeatherData getCurrentWeather(String location){ 
+           
+        WeatherAPIKey__mdt weaherApiKey = [SELECT Key__c FROM WeatherAPIKey__mdt LIMIT 1];  
```
please use unique Label filter in SOQL

okrasko55 commented on force-app/main/default/classes/WeatherService.cls:
```diff
@@ -0,0 +1,33 @@
+        Http http = new Http();
+        HttpResponse resp = http.send(req);
+       
+        Map<String, Object> weatherMap = (Map<String, Object>) JSON.deserializeUntyped(resp.getBody());
```
don't use Hungarian notation

okrasko55 replied:
one more thing is it needs to check if response is success first and after that work with body of it

okrasko55 commented on force-app/main/default/classes/WeatherService.cls:
```diff
@@ -0,0 +1,33 @@
+    public static WeatherData getCurrentWeather(String location){ 
+           
+        WeatherAPIKey__mdt weaherApiKey = [SELECT Key__c FROM WeatherAPIKey__mdt LIMIT 1];  
+        String apiKey = weaherApiKey.Key__c;
```
check if weaherApiKey exists first and after that take value from property, in this particular case Key__c

okrasko55 commented on force-app/main/default/classes/WeatherService.cls:
```diff
@@ -0,0 +1,33 @@
+        String apiKey = weaherApiKey.Key__c;
+
+        HttpRequest req = new HttpRequest();
+        req.setEndpoint('callout:WeatherAPI?q='+ location + '&units=metric' + '&appid=' + apiKey);
```
potentially apiKey might be null, what response service returns in this case?

okrasko55 commented:
Test class is missing

InnaKamenskaya committed:
Fixed code review bugs. Tests has been added

InnaKamenskaya committed:
metaData apiKey check for null has been added

okrasko55 commented on force-app/main/default/classes/TestWeatherService.cls:
```diff
@@ -0,0 +1,24 @@
+@isTest
+private class TestWeatherService {
+    private static String location = 'Los Angeles';
```
please make it final and all letters in capital since it is constant

okrasko55 commented on force-app/main/default/classes/WeatherAPICalloutMock.cls:
```diff
@@ -0,0 +1,27 @@
+
+    public WeatherAPICalloutMock(String location) {
+        if (location == null || location == '') {
+            weather = null;
```
it's not clear what location parameter for? it is not being used later in a code.

okrasko55 commented on force-app/main/default/classes/WeatherAPICalloutMock.cls:
```diff
@@ -0,0 +1,27 @@
+        if (location == null || location == '') {
+            weather = null;
+        }else { 
+            weather = new WeatherService.WeatherData();
+            weather.temp = 25.5;
+            weather.windSpeed = 3.3;
+            weather.humidity = 75.0;
```
I would implement it by the way that constructor accepts WeatherService.WeatherData instance and then assign it to weather variable from 3th line, thus it will be more dynamic without hardcoding

okrasko55 commented on force-app/main/default/classes/WeatherService.cls:
```diff
@@ -0,0 +1,45 @@
+            return null;
+        }
+        HttpRequest req = new HttpRequest();
+        req.setEndpoint('callout:WeatherAPI?q='+ location + '&units=metric' + '&appid=' + apiKey);
```
would be nice to use String.format() method here

okrasko55 commented on force-app/main/default/lwc/weatherComponent/weatherComponent.js:
```diff
@@ -0,0 +1,89 @@
+                });
+            }            
+            loadStyle(this, weather);
+        }else if(result.error){
```
am not sure "else if" is require here, I mean just "else" should be enough, if result.data not equals true then result.error should equal true, please correct me if I am wrong

InnaKamenskaya replied:
I need to get access to the error that the result variable provides and then display the specific message of this result.error, that's why I use the "else if" statement

InnaKamenskaya committed:
Variable "LOCATION" in TestWeatherService class was changed to constant;
constructor of the WeatherAPICalloutMock class was changed to accepts WeatherService.WeatherData instance;
String.format() was added to WeatherService class for making endpoint string more dynamical

Zerony commented on force-app/main/default/classes/WeatherAPICalloutMock.cls:
```diff
@@ -0,0 +1,20 @@
+        HttpResponse response = new HttpResponse();
+        response.setHeader('Content-Type', 'application/json');
+        if (weather == null) {
+            response.setStatusCode(400);
```
minor. Good practice to return value immediately, to avoid else statemets.

Zerony commented on force-app/main/default/classes/WeatherService.cls:
```diff
@@ -0,0 +1,47 @@
+            return null;
+        }
+        String apiKey = ([SELECT Key__c FROM WeatherAPIKey__mdt WHERE Label =: META_DATA_LABEL] == null) ? null : [SELECT Key__c FROM WeatherAPIKey__mdt WHERE Label =: META_DATA_LABEL].Key__c;
+        if (apiKey == null || apiKey == '') {
```
Better to use String.isBlank();

InnaKamenskaya replied:
Fixed

Zerony commented on force-app/main/default/classes/WeatherService.cls:
```diff
@@ -0,0 +1,47 @@
+global class WeatherService {
```
Bad class name. Better WeatherRepository

Zerony commented on force-app/main/default/classes/WeatherService.cls:
```diff
@@ -0,0 +1,47 @@
+
+        WeatherData weather = new WeatherData();
+        if(Test.isRunningTest()){
+            weather = (WeatherData)JSON.deserialize(resp.getBody(), WeatherData.class);
```
same here. Better to return immediately.

InnaKamenskaya replied:
Fixed

Zerony commented on force-app/main/default/classes/WeatherService.cls:
```diff
@@ -0,0 +1,47 @@
+        if (resp == null) {
+            return null;
+        }
+        Map<String, Object> weatherMap = (Map<String, Object>) JSON.deserializeUntyped(resp.getBody());
```
Would be good to build model class for response.

InnaKamenskaya replied:
model class for response has been changed.

Zerony commented on force-app/main/default/classes/WeatherService.cls:
```diff
@@ -0,0 +1,47 @@
+        Http http = new Http();
+        HttpResponse resp = http.send(req);
+        if (resp == null) {
+            return null;
```
Silent fail. Better to throw exception.

InnaKamenskaya replied:
Fixed

Zerony commented on force-app/main/default/lwc/weatherComponent/weatherComponent.js:
```diff
@@ -0,0 +1,89 @@
+    }   
+
+    get currentTemp(){
+        return Math.round(this.weather.temp) +  '\u2103';
```
Magic constant. '\u2103' should be named.

InnaKamenskaya replied:
Fixed

Zerony commented on force-app/main/default/classes/WeatherService.cls:
```diff
@@ -0,0 +1,47 @@
+        if(location == null || location ==''){
+            return null;
+        }
+        String apiKey = ([SELECT Key__c FROM WeatherAPIKey__mdt WHERE Label =: META_DATA_LABEL] == null) ? null : [SELECT Key__c FROM WeatherAPIKey__mdt WHERE Label =: META_DATA_LABEL].Key__c;
```
would be good to extract to method.

InnaKamenskaya replied:
Fixed

Zerony commented on force-app/main/default/classes/WeatherService.cls:
```diff
@@ -0,0 +1,47 @@
+
+    @AuraEnabled(cacheable=true)
+    public static WeatherData getCurrentWeather(String location){ 
+        if(location == null || location ==''){
```
Better to use String.isBlank

InnaKamenskaya replied:
Fixed

InnaKamenskaya committed:
change weather API key type from Custom MetaData to Custom Settings.
change tests for WeatherRepository to test with Custom Settings.

InnaKamenskaya committed:
Model class for response has been changed to simplify JSON deserializing.
Tests was moved to the tests folder
